sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -98,6 +99,21 @@ object DateTimeUtils {
     sdf
   }
 
+  private val threadLocalTimeZones = new ThreadLocal[mutable.Map[String, TimeZone]] {
+    override def initialValue(): mutable.Map[String, TimeZone] = mutable.Map.empty
+  }
+
+  def getTimeZone(timeZoneId: String): TimeZone = {
+    val timeZones = threadLocalTimeZones.get()
```

Comment:
How about just `threadLocalTimeZones.get().getOrElseUpdate(timeZoneId, TimeZone.getTimeZone(timeZoneId))`? It avoids the double lookup.